src/parser/parser.lalrpop

```diff
@@ -159,9 +160,9 @@ RootExpr: Node<RootExpr> = {
 // -----------------------------------------------------------------------------
 
 #[inline]
-Exprs: Vec<Node<Expr>> = {
-    Expr => vec![<>],
-    <v:(<Expr> EndOfExpression)+> <e:(<Expr>)?> => match e {
+Exprs<T>: Vec<Node<Expr>> = {
```

Comment:
It seems like this change required a lot of changes to the parser. Is this because you introduced `ExprWithoutReturn`?

Reply:
Yes, exactly. Basically, if we want the parser to already refuse the code, we need the parser to know whether it is inside a closure, leading to two sets of `Exprs`, `Block`s and so on.

The other two options are checking the AST after parsing in another pass (what was here previously) or doing this bottom-up probably during `compile` where the `Expr` struct (and probably others, I'm not sure) would have an added field with the information of whether there is a return inside and failing the compilation when a closure finds out it contains a return.

Or a decision could be made how the returns should behave inside the closure. Then none of these would be needed. Any of these three options would add code and complexity that would be ultimately deleted when returns in closures are allowed.

Reply:
Based on this https://github.com/vectordotdev/vector/pull/19828/files#r1488734971, do you think it is viable to implement option 2?

Reply:
Yeah, I think so.

Reply:
Thank you, I think this is the desirable path forward.

Reply:
Hi @mladedav, just double checking, is this something you plan to implement in this PR?

Reply:
Yes, I'm still planing to. It might be a bit more difficult than I originally thought because the closure must be guaranteed to only return correct types (e.g. bool for `filter`) but I didn't have enough time to deep dive into this yet.

Reply:
It seems to be working now, if you have any feedback, it would be appreciated.

I'm somewhat uneasy with the changes because I think I might have very easily overlooked a place where the returns should be propagated.